### ISSUE TITLE
Template generation does not suggest enabling MultiParamTypeClasses (lts-16.5, persistent-template-2.8.2.3)

### ISSUE THREAD

**Issue by BorisNikulin** [yesodweb/persistent#1105]
TL;DR template generation picks up on missing extensions and suggests enabling them except for `MultiParamTypeClasses`.
Related to #1093.

While setting up persistent et al. for the first time, template generation failed a couple times.

<details>
<summary>
Gripe about error messages when putting Maybe before the data type.
</summary>
First, due to putting Maybe before the type as in `field Maybe Text` as that's my Haskell syntax intuition (was not expecting to use SQL style). The error message is not great or at all helpful. Only mentioning a kind mismatch. I do understand fancy type trickery/TH can yield less than desirable error messages, but, a concise error message would be helpful.
</details>

Second, I started off the file with only `-XTemplateHaskell` and `-XQuasiQuotes`. The quasi quoter very helpfully pointed out I needed `TypeFamilies` and `GADTs`.
After adding those it also pointed out I needed `DerivingStrategies` `GenaralizedNewtypeDeriving`, `StandAloneDeriving`, and `UndecidableInstances`. Another compile later and I got multiple multiparam typeclass errors stating that instances where illegal due to only one type allowed in a head. Example of one below:
```
    • Illegal instance declaration for
        ‘persistent-2.10.5.2:Database.Persist.Class.PersistStore.ToBackendKey
           Database.Persist.Sql.Types.Internal.SqlBackend Event’
        (Only one type can be given in an instance head.
         Use MultiParamTypeClasses if you want to allow more, or zero.)
    • In the instance declaration for
        ‘persistent-2.10.5.2:Database.Persist.Class.PersistStore.ToBackendKey Database.Persist.Sql.Types.Internal.SqlBackend Event’
```
This error by GHC is pretty helpful and is probably sufficient for pointing out to enable `MultiParamTypeClasses` though considering other extensions are suggested, this one could also be suggested.

**Comment by BorisNikulin** [comment]
Relevant lines include:
https://github.com/yesodweb/persistent/blob/master/persistent-template/Database/Persist/TH.hs#L1898
https://github.com/yesodweb/persistent/blob/master/persistent-template/Database/Persist/TH.hs#L415

Unsure if the required extension should be a part of the base set of extensions, though the typeclass being named ToBackendKey suggests the extension may be a core requirement and should be added to the base set of required extensions (second link at line 415).

Also, it seem like all the linked above required extensions should be in one set instead of 2 different sets that activate after the first one is satisfied which requires a compile, a fix, and then another compile, and finally a fix.

**Comment by MaxGabriel** [comment]
So, I'm not super deep on `MultiParamTypeClasses`, but my understanding is it's a very standard typeclass. I think with the right combination of parameters to `mkKeyTypeDec` you _might_ be able to avoid it (it's kinda hard to read the code), but I'm fine always requiring it in `requirePersistentExtensions`.

**Comment by BorisNikulin** [comment]
`ToBackendKey` is created in the following lines
https://github.com/yesodweb/persistent/blob/be8901eac2714e09a083c0d46bb9a2073650c497/persistent-template/Database/Persist/TH.hs#L908-L918
and does require `MultiParamTypeclasses`.

The function that runs that code
https://github.com/yesodweb/persistent/blob/be8901eac2714e09a083c0d46bb9a2073650c497/persistent-template/Database/Persist/TH.hs#L855-L891
uses those decls on lines
https://github.com/yesodweb/persistent/blob/be8901eac2714e09a083c0d46bb9a2073650c497/persistent-template/Database/Persist/TH.hs#L874
and
https://github.com/yesodweb/persistent/blob/be8901eac2714e09a083c0d46bb9a2073650c497/persistent-template/Database/Persist/TH.hs#L864
via the function `genericNewtypeInstances` which conditionally uses `backendKeyGenericI` as followed
https://github.com/yesodweb/persistent/blob/be8901eac2714e09a083c0d46bb9a2073650c497/persistent-template/Database/Persist/TH.hs#L938-L939

The `mkKeyTypeDec` function overall requires the base set of extensions, but the actual requirement of `MultiParamTypeclasses` is conditional on some logic
```
(not $ mpsGeneric mps) && useNewType
```
for `backendKeyGenericI` or
```
mpsGeneric mps && useNewType && (not (defaultIdType t) || not useNewtype || isJust (entityPrimary t))
```
for `genericNewtypeInstances`.

So it seems that one can conservatively add `MulitParamTypeclasses` to the base set or in those specific branches put in an extra typeclass requirement.

**Comment by parsonsmatt** [comment]
This should be fixed in 2.11.
